Add "list" command to show all available stacks

kainosnoema opened the pull request:
An open question is whether this command should show stacks already in CF (like show does), or just the local stacks. Or maybe both, with status of creation? Currently this just lists local stacks.

![screen shot 2018-08-02 at 11 26 01 am](https://user-images.githubusercontent.com/145951/43603119-e59b92e8-9646-11e8-9792-81dd68f9f1ec.png)
![screen shot 2018-08-02 at 3 45 42 pm](https://user-images.githubusercontent.com/145951/43615146-26d8fc18-966b-11e8-85f6-cc0444be2328.png)

Commit:
Add "list" command to show all available stacks

Signed-off-by: Evan Owen <evan@zinc.it>

kainosnoema commented on backend/config.go:
```diff
@@ -86,6 +95,14 @@ func (s *configStore) resolveStack(path string, st stackConfig) stackConfig {
 		}
 
+		if stack.Parameters == nil {
+			stack.Parameters = map[string]interface{}{}
```
Fixes panic when Parameters aren't defined.

kainosnoema commented on backend/config.go:
```diff
@@ -86,6 +95,14 @@ func (s *configStore) resolveStack(path string, st stackConfig) stackConfig {
+		}
+
+		if stack.TemplateName == "" {
+			stack.TemplateName = stack.Name
```
This is documented in the readme, but wasn't working....

defnotpaul commented on backend/fetcher.go:
```diff
@@ -38,15 +38,28 @@ func newFetcher(cs ConfigStore, ts TemplateStore, r ParamsResolver) *fetcher {
+	return f.fetchTemplates(stackConfigs)
+}
+
 func (f *fetcher) Fetch(name string) ([]client.Stack, error) {
```
Should we maybe just update these to return a `client.StackList`?

kainosnoema replied:
Yes probably. Didn’t know if it was a CLI concern or not but doesn’t hurt.

defnotpaul replied:
Hmmm, yeah, that's a good point. I think for now it makes sense, if only for the convenience.

Really though, I think we probably should be returning concrete types here.

eyeamera commented:
I dig this. I was thinking about implementing a `list` feature but got hung up on the same questions you're having. I think supplying a flag for `--remote` or `--local` might make sense if we want to support both?

kainosnoema commented:
@eyeamera I’ve made some changes locally to support showing remote stacks and their status too. Adds a bit of code that I want to clean up though. Thoughts?

kainosnoema commented:
Ah missed your comment. Sounds good, I’ll work something up.

Commit:
Move common interfaces into common module for backend + client to share

This helps clear up confusion about which package should own the types
and reduces duplication in the backend types.

Signed-off-by: Evan Owen <evan@zinc.it>

kainosnoema commented:
@eyeamera ok this is updated to use a string option and I've moved the common types into their own package.

eyeamera commented:
Sweet, this is awesome!
